internal/telemetry/metrics.go

```diff
@@ -88,21 +104,59 @@ func (c *collector) Describe(ch chan<- *prometheus.Desc) {
 }
 
 func (c *collector) Collect(ch chan<- prometheus.Metric) {
-	ch <- prometheus.MustNewConstMetric(c.infoDesc, prometheus.GaugeValue, 1)
-	ch <- prometheus.MustNewConstMetric(c.objectDefsDesc, prometheus.GaugeValue, countObjectDefs(c.ds))
-}
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
 
-func countObjectDefs(ds datastore.Datastore) float64 {
-	ctx := context.Background()
-	rev, err := ds.OptimizedRevision(ctx)
+	dsStats, err := c.ds.Statistics(ctx)
 	if err != nil {
-		return -1
+		log.Warn().Err(err).Msg("unable to collect datastore statistics")
 	}
 
-	namespaces, err := ds.ListNamespaces(ctx, rev)
-	if err != nil {
-		return -1
-	}
+	ch <- prometheus.MustNewConstMetric(c.infoDesc, prometheus.GaugeValue, 1)
+	ch <- prometheus.MustNewConstMetric(c.objectDefsDesc, prometheus.GaugeValue, float64(len(dsStats.ObjectTypeStatistics)))
+	ch <- prometheus.MustNewConstMetric(c.relationshipsDesc, prometheus.GaugeValue, float64(dsStats.EstimatedRelationshipCount))
 
-	return float64(len(namespaces))
+	dispatchedCountMetrics := make(chan prometheus.Metric)
+	g := errgroup.Group{}
+	g.Go(func() error {
+		for {
```

Comment:
I think this is one of those rare cases where you can actually range over the channel

Reply:
Neat, I didn't even know this was a thing!